source/matrix_free/vector_data_exchange.cc

```diff
@@ -316,6 +318,965 @@ namespace internal
             ghost_array[j] = 0.;
       }
 
+
+
+      namespace internal
+      {
+        void
+        compress(std::vector<unsigned int> &recv_sm_ptr,
+                 std::vector<unsigned int> &recv_sm_indices,
+                 std::vector<unsigned int> &recv_sm_len)
+        {
+          std::vector<unsigned int> recv_ptr = {0};
+          std::vector<unsigned int> recv_indices;
+          std::vector<unsigned int> recv_len;
+
+          for (unsigned int i = 0; i + 1 < recv_sm_ptr.size(); i++)
+            {
+              if (recv_sm_ptr[i] != recv_sm_ptr[i + 1])
+                {
+                  recv_indices.push_back(recv_sm_indices[recv_sm_ptr[i]]);
+                  recv_len.push_back(1);
+
+                  for (unsigned int j = recv_sm_ptr[i] + 1;
+                       j < recv_sm_ptr[i + 1];
+                       j++)
+                    if (recv_indices.back() + recv_len.back() !=
+                        recv_sm_indices[j])
+                      {
+                        recv_indices.push_back(recv_sm_indices[j]);
+                        recv_len.push_back(1);
+                      }
+                    else
+                      recv_len.back()++;
+                }
+              recv_ptr.push_back(recv_indices.size());
+            }
+
+          recv_sm_ptr = recv_ptr;
+          recv_sm_ptr.shrink_to_fit();
+          recv_sm_indices = recv_indices;
+          recv_sm_indices.shrink_to_fit();
+          recv_sm_len = recv_len;
+          recv_sm_len.shrink_to_fit();
+        }
+
+        std::vector<unsigned int>
+        procs_of_sm(const MPI_Comm &comm, const MPI_Comm &comm_shared)
+        {
+          std::vector<unsigned int> ranks_shared;
+
+#ifndef DEAL_II_WITH_MPI
+          Assert(false, ExcNeedsMPI());
+          (void)comm;
+          (void)comm_shared;
+#else
+          // extract information from comm
+          int rank_;
+          MPI_Comm_rank(comm, &rank_);
+
+          const unsigned int rank = rank_;
+
+          // extract information from sm-comm
+          int size_shared;
+          MPI_Comm_size(comm_shared, &size_shared);
+
+          // gather ranks
+          ranks_shared.resize(size_shared);
+          MPI_Allgather(&rank,
+                        1,
+                        MPI_UNSIGNED,
+                        ranks_shared.data(),
+                        1,
+                        MPI_INT,
+                        comm_shared);
```

Comment:
To expand on the previous comment, for the current arguments the correct way would be
```suggestion
          MPI_Allgather(&rank,
                        1,
                        MPI_UNSIGNED,
                        ranks_shared.data(),
                        1,
                        MPI_UNSIGNED,
                        comm_shared);
```
But I would change the first to `MPI_INT`.